src/entities.lisp

```diff
@@ -0,0 +1,96 @@
+;;;; entities.lisp
+
+(in-package #:sketch)
+
+;;;  _____ _   _ _____ ___ _____ ___ _____ ____
+;;; | ____| \ | |_   _|_ _|_   _|_ _| ____/ ___|
+;;; |  _| |  \| | | |  | |  | |  | ||  _| \___ \
+;;; | |___| |\  | | |  | |  | |  | || |___ ___) |
+;;; |_____|_| \_| |_| |___| |_| |___|_____|____/
+;;;
+
+;;; TODO:
+;;; - Better code reuse with sketch.lisp
+
+(defparameter *entity* nil
+  "The current entity.")
+
+(defclass entity ()
+  ((width :initform 100 :accessor entity-width :initarg :width)
+   (height :initform 100 :accessor entity-height :initarg :height)))
+
+(defmethod register-entity ((sketch sketch) (entity entity) box)
+  (setf (gethash entity (sketch-%entities sketch))
+	;; TODO: sb-cga shouldn't be used directly from here.
+	(cons (sb-cga:inverse-matrix (env-model-matrix (sketch-%env sketch))) box)))
+
+(defmethod initialize-instance :after ((instance entity) &rest initargs &key &allow-other-keys)
+  (apply #'prepare instance initargs))
+
+(defmethod update-instance-for-redefined-class :after
+    ((instance entity) added-slots discarded-slots property-list &rest initargs)
+  (declare (ignore added-slots discarded-slots property-list))
+  (apply #'prepare instance initargs))
+
+(defun define-entity-defclass (name bindings)
+  `(defclass ,name (entity)
+     (,@(loop for b in bindings
+              unless (eq 'entity (binding-prefix b))
+		collect `(,(binding-name b)
+                          :initarg ,(binding-initarg b)
+                          :accessor ,(binding-accessor b)
+			  ,@(when (binding-channelp b) '(:allocation :class)))))))
```

Comment:
Why making channel bindings same for all instances?

IMHO it would be more useful to make channels more sketch-local and also entity-local, so that you could have different values for separate entities.

For example take a button that should be hovered: you could use a channel here.
```lisp
(defentity button ((hovered nil :channel-name :hover) ...)
  (with-pen (make-pen :fill (if hovered hovered-color usual-color))
    ...))

(defmethod on-enter ((entity button))
  (out :hover t))

(defmethod on-leave ((entity button))
  (out :hover nil))

Reply:
How would that work if you are trying to send something between sketches, from the repl, or from wherever else?

Reply:
From REPL you could say
```common-lisp
(with-sketch-channels (<uses-saved-sketch-instance>)
  (out :chan-name value))
```
or having `(kit.sdl2:last-window)` instead of saved sketch instance by default (although that might be a bit tricky to do without run-time overhead for `out` itself -- maybe a separate macro `out*` or `send` that would use last sketch automatically.
```common-lisp
(send <name> <value>)
;; expands to
(with-sketch-channels ((kit.sdl2:last-window))
  (out <name> <value>))
```

To send something between sketches you'd need to store another sketch somewhere and simply use `with-sketch-channels` before sending (or, maybe `send` could also take an optional parameter which defaults to `(kit.sdl2:last-window)`, so it could be defined like this:
```common-lisp
(defmacro send (name value &optional (sketch '(kit.sdl2:last-window)))
  `(with-sketch-channels (,sketch)
     (out ,name ,value)))
```

Entities would need to have an environment of channels that has a parent (top-level sketch or parenting entity), so a bit more complicated setup than a hash-table might be needed (not too complicated though IMO). One possible feature needed is to be able to tell apart cases where you set an "entity local" channel or when you want to set parent's channel.